src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Taken from https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/Formatters/TranscodingWriteStream.cs
+
+using System.Buffers;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    internal sealed class TranscodingWriteStream : Stream
+    {
+        internal const int MaxCharBufferSize = 4096;
+        internal const int MaxByteBufferSize = 4 * MaxCharBufferSize;
+        private readonly int _maxByteBufferSize;
+
+        private readonly Stream _stream;
+        private readonly Decoder _decoder;
+        private readonly Encoder _encoder;
+        private char[] _charBuffer;
+        private int _charsDecoded;
+        private bool _disposed;
+
+        public TranscodingWriteStream(Stream stream, Encoding targetEncoding)
+        {
+            _stream = stream;
+
+            _charBuffer = ArrayPool<char>.Shared.Rent(MaxCharBufferSize);
+
+            // Attempt to allocate a byte buffer than can tolerate the worst-case scenario for this
+            // encoding. This would allow the char -> byte conversion to complete in a single call.
+            // However limit the buffer size to prevent an encoding that has a very poor worst-case scenario.
+            _maxByteBufferSize = Math.Min(MaxByteBufferSize, targetEncoding.GetMaxByteCount(MaxCharBufferSize));
+
+            _decoder = Encoding.UTF8.GetDecoder();
+            _encoder = targetEncoding.GetEncoder();
+        }
+
+        public override bool CanRead => false;
+        public override bool CanSeek => false;
+        public override bool CanWrite => true;
+        public override long Length => throw new NotSupportedException();
+        public override long Position { get; set; }
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+            => _stream.FlushAsync(cancellationToken);
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
```

Comment:
Nit: some of the methods are using `=> throw ...`, and some are using `{ throw ... }`.  We should be consistent.